webdriver/support/fixtures.py

```diff
@@ -83,54 +88,88 @@ def create_window():
-        request.addfinalizer(destroy)
-        request.addfinalizer(lambda: _switch_to_top_level_browsing_context(session))
-        request.addfinalizer(lambda: _restore_windows(session))
-        request.addfinalizer(lambda: _dismiss_user_prompts(session))
-        request.addfinalizer(lambda: _ensure_valid_window(session))
-
-        return session
+
+def configuration():
+    host = os.environ.get("WD_HOST", default_host)
+    port = int(os.environ.get("WD_PORT", default_port))
+    capabilities = json.loads(os.environ.get("WD_CAPABILITIES", "{}"))
+
+    return {
+        "host": host,
+        "port": port,
+        "capabilities": capabilities
+    }
+
+
+_current_session = None
+
+
+def session(configuration, request):
+    """Create and start a session for a test that does not itself test session creation.
+
+    By default the session will stay open after each test, but we always try to start a
+    new one and assume that if that fails there is already a valid session. This makes it
+    possible to recover from some errors that might leave the session in a bad state, but
+    does not demand that we start a new session per test."""
+    global _current_session
+    if _current_session is None:
+        _current_session = webdriver.Session(configuration["host"],
+                                             configuration["port"],
+                                             capabilities=configuration["capabilities"])
+    try:
+        _current_session.start()
+    except webdriver.errors.SessionNotCreatedException:
+        if not _current_session.session_id:
+            raise
+
+    # finalisers are popped off a stack,
+    # making their ordering reverse
+    request.addfinalizer(lambda: _switch_to_top_level_browsing_context(_current_session))
+    request.addfinalizer(lambda: _restore_windows(_current_session))
+    request.addfinalizer(lambda: _dismiss_user_prompts(_current_session))
+    request.addfinalizer(lambda: _ensure_valid_window(_current_session))
+
+    return _current_session
+
+
+def new_session(configuration, request):
+    """Return a factory function that will attempt to start a session with a given body.
+
+    This is intended for tests that are themselves testing new session creation, and the
+    session created is closed at the end of the test."""
+    def end():
+        global _current_session
+        if _current_session is not None and _current_session.session_id:
+            _current_session.end()
+            _current_session = None
+
+    def create_session(body):
+        global _current_session
+        _session = webdriver.Session(configuration["host"],
+                                     configuration["port"],
+                                     capabilities=None)
+        # TODO: merge in some capabilities from the confguration capabilities
+        # since these might be needed to start the browser
+        value = _session.send_command("POST", "session", body=body)
```

Comment:
What would you think about simply constructing the Session object (with
capabilities) and returning it? This would make the tests that need this
fixture a little more direct, and it would save us from having to deal with the
command return value or the object's `session_id` property.

Also, why is the "merging" operation a "TODO".